src/result.rs

```diff
@@ -1,35 +1,77 @@
+use std::{io, fmt};
 
 /// A unified Result type representing the outcome of method calls that may fail.
 pub type IonResult<T> = Result<T, IonError>;
 
 /// Represents the different types of high-level failures that might occur when reading Ion data.
-#[derive(Debug, Fail, Clone, PartialEq)]
+#[derive(Debug, Error)]
 pub enum IonError {
     /// Indicates that an IO error was encountered while reading or writing.
-    #[fail(display = "An IO error occurred: {}", description)]
-    IoError { description: String },
+    #[error("{source:?}")]
+    IoError {
+        #[from]
+        source: io::Error,
+    },
+
+    #[error("{source:?}")]
+    FmtError {
+        #[from]
+        source: fmt::Error,
+    },
 
     /// Indicates that the data stream being read contained illegal or otherwise unreadable data.
-    #[fail(display = "A decoding error occurred: {}", description)]
+    #[error("{description}")]
     DecodingError { description: String },
 
     /// Returned when the user has performed an illegal operation (for example: calling stepOut()
     /// on the cursor at the top level.)
-    #[fail(
-        display = "The user has performed an action that is not legal in the current state: {}",
-        description
-    )]
-    IllegalOperation { description: String },
+    #[error("The user has performed an operation that is not legal in the current state: {operation}")]
+    IllegalOperation { operation: String },
 }
 
-/// A convenience method for creating an IonResult containing an IonError::IoError with the
-/// provided description text.
-pub fn io_error<T>(description: &str) -> IonResult<T> {
-    Err(IonError::IoError {
-        description: description.to_string(),
-    })
+// io::Error does not implement Clone, which precludes us from simply deriving an implementation.
+// IonError needs a Clone implementation because we use a jump table of cached IonResult values when
+// parsing type descriptor bytes. The only error type that will be cloned by virtue of using the jump
+// table is DecodingError.
+impl Clone for IonError {
+    fn clone(&self) -> Self {
+        use IonError::*;
+        match self {
+            IoError {source} => IoError {
+                // io::Error implements From<ErrorKind>, and ErrorKind is cloneable.
+                source: io::Error::from(source.kind().clone())
+            },
+            FmtError {source} => FmtError {
+                source: source.clone()
+            },
+            DecodingError {description} => DecodingError {
+                description: description.clone()
+            },
+            IllegalOperation {operation} => IllegalOperation {
+                operation: operation.clone()
+            }
+        }
+    }
+}
+
+// io::Error does not implement PartialEq, which precludes us from simply deriving an implementation.
+// Providing an implementation of PartialEq allows IonResult values to be the left or right side in
+// an assert_eq!() statement.
+impl PartialEq for IonError {
```

Comment:
It looks like we ought to implement `Eq` as well. 
https://www.philipdaniels.com/blog/2019/rust-equality-and-ordering/